### ISSUE TITLE
GSPO leaks masked tokens through batch centering and is not sequence-level

### ISSUE THREAD

**Issue by Amix29** [MakazhanAlpamys/Soup#723]
## Summary

Soup's `gspo` branch centers each token log-ratio across the batch before applying
the completion mask:

```python
log_ratio_centered = log_ratio - log_ratio.mean(dim=0, keepdim=True)
ratio_stab = torch.exp(log_ratio_centered)
token_loss = -(ratio_stab * advantages_2d)
return _masked_mean(token_loss, completion_mask)
```

Consequently, a masked token participates in the batch mean, changes the loss of
unmasked examples, and receives a non-zero gradient. The implementation also does
not match the published GSPO objective, which uses a length-normalized sequence
likelihood ratio and sequence-level clipping rather than per-column batch
centering: https://qwenlm.github.io/blog/gspo/

## Reproduction

Reproduced against current `main` at
`9f1e48b33a5fcda7e621e612ec9305cb52a38e07` with Python 3.12.2 and PyTorch
2.13.0, executing the actual `apply_variant_loss()` source.

In a 4x2 synthetic batch, changing only one masked log-ratio gives:

| Masked log-ratio | Loss | Gradient at masked token |
|---:|---:|---:|
| `-2` | `-0.3333333` | `0.0833333` |
| `-6` | `-0.9060939` | `0.2265235` |

DAPO, BNPO, and `two_sided` controls are invariant to the same mutation and return
zero gradient at that masked position.

## Impact

Selecting `grpo_variant: gspo` optimizes an objective influenced by padding or
otherwise masked completion positions. More broadly, the named variant is not the
sequence-level GSPO algorithm users reasonably expect.

This is distinct from #123 (initial kernel implementation), #584 (the variants
previously falling back to stock TRL), and #342 (non-finite gradients in the
standard `variant=None` path).

## Suggested fix

- Exclude masked positions before every statistic or reduction.
- Compute the masked, length-normalized log-probability ratio per completion.
- Apply GSPO clipping and the policy-gradient surrogate at sequence level according
  to the published definition.
- Preserve any KL term and normalization expected by the surrounding TRL contract.

## Acceptance criteria

- [ ] Changing a masked token cannot change the GSPO loss.
- [ ] Every masked token receives exactly zero gradient.
- [ ] The implementation matches a direct reference calculation of the published
      sequence-level objective, including variable completion lengths.
- [ ] Tests cover batch permutation, padding position, all-masked rejection, clip
      boundaries, and unequal sequence lengths.
- [ ] Existing runtime integration coverage proves the corrected branch is reached
      during an actual Trainer loss call.

**Comment by MakazhanAlpamys** [comment]
**Half of this shipped in `e1f5f69` by @AmirF194 in #735. This issue stays OPEN for the other half.**

## What is fixed

The masked-token leak through batch centering. `gspo` centred the importance log-ratio with a plain `log_ratio.mean(dim=0)` over the whole tensor, padding positions included, so every unmasked token's loss and gradient moved with values belonging to no completion. Measured on a 4x2 batch, varying only a masked position's log-ratio:

```
masked log-ratio    before        after
     -2.0        -0.33205155   -0.28571433
     -6.0        -0.53440636   -0.28571433
    +50.0        -0.21428600   -0.28571433
```

Gradient at the masked position: `0.0800301731` -> exactly `0.0`. With an all-ones mask the loss is **bitwise identical** to before, so nobody whose batches do not pad sees a change, and the other six variants are unchanged to the last digit.

## What is NOT fixed, and why the issue stays open

The second claim: **`gspo` is not the published sequence-level objective.** Nothing in #735 touches that, and @AmirF194 said so in the PR body before being asked rather than taking the issue number and leaving me to find the gap:

> I didn't touch the sequence-level part... that reads like its own PR once you've said what clip bounds you want

That is the right call and the question is genuinely mine to answer, not his. So, on the record, the open work is:

1. **The sequence-level objective itself** — the ratio and the clip applied per sequence rather than per token.
2. **The clip bounds**, which is the decision the above is waiting on and which I have not made.
3. **The tests #723 asks for that #735 does not provide**: invariance under batch permutation, behaviour at the clip boundaries, and unequal sequence lengths.
4. **A real-`GRPOTrainer` runtime proof that the gspo branch is reached.** There is a mock-TRL contract test covering it (`test_v05311.py:676`, `:727`), but the live leg is red on my box from the #651 torch/trl floor, so it is pinned by a double rather than by a run.

One thing worth stating so it is not rediscovered as a finding: `docs/training.md:535` describes gspo as a "group-stabilised importance ratio with per-batch control variate", which is honest — the docs never claimed the paper's algorithm. It is the variant **name** that implies it, and that is item 1's problem, not a documentation bug.

`help wanted` stays on. If you take item 1, say so here and I will label it `claimed` for you the same day.

**Comment by kok-o** [comment]
Hi @MakazhanAlpamys, I would like to take on the remaining open work (Item 1: the sequence-level GSPO objective and clipping).

My plan:
1. Implement the sequence-level, length-normalized log-likelihood ratio and sequence-level clipping according to the published GSPO definition, ensuring masked positions are excluded before statistics/reductions.
2. Add the requested test suite: invariance under batch permutations, behavior at clip boundaries, unequal sequence lengths, and padding masks.
3. Validate runtime integration so the corrected branch is reached cleanly during training.

Could you please mark this issue as `claimed` for me? Also, please let me know if you already have specific clip bounds in mind, or if you'd prefer starting with standard symmetric bounds.

**Comment by MakazhanAlpamys** [comment]
Closed by #744, merged as `0139553d`, thanks to @kok-o.

**One correction to this issue's own diagnosis, because it matters for anyone reading it later.** The masked-token leak this issue is titled after had already been closed by #735 — masked gradients are zero on `main` before #744. What #735 left behind is the defect #744 actually fixes: column centering zeroes the **policy** gradient on any column with a single unmasked member, so real unmasked positions were learning nothing (2 of 9 in an ordinary fixture). The "not sequence-level" half of the title was entirely correct and is now closed properly.

`grpo_variant: gspo` is now the published GSPO objective (Qwen Team, arXiv:2507.18071): a length-normalized sequence importance ratio `s_i = exp((1/|y_i|) * sum_t (log p_new - log p_old))` with sequence-level surrogate clipping.

**Breaking for reproducibility:** existing configs using `grpo_variant: gspo` will not reproduce prior losses or gradients. Stated in the changelog fragment.

The one suggested-fix bullet not addressed is "preserve any KL term" — `beta` is ignored by `gspo`. I checked, and it is ignored by `dapo`, `bnpo`, `dr_grpo`, `rft` and `two_sided` as well, so that is a separate project-wide question rather than something this PR skipped.

Try:

```
soup train  # task='grpo' + training.grpo_variant=gspo [training.grpo_delta=0.0003]
```
